### ISSUE TITLE
CRAN version of packrat (0.4.7-1) not compatible with latest version of R (3.3.0)

### ISSUE THREAD

**Issue by grrothman** [rstudio/packrat#296]
It appears that packrat is not yet up to date with the newest version of R. When activating packrat in a new RStudio project running R v3.3.0, the RStudio instance freezes up as soon as any source code is opened (either by creating a new script with cmd+shift+N or opening an existing file with cmd+O), and this issue is resolved by running a legacy version of R (v3.2.4 or earlier).

**Comment by kevinushey** [comment]
I wasn't able to replicate this. Can you provide some more information? What version of RStudio are you running with, and what operating system?

**Comment by kevinushey** [comment]
Is it possible that packrat is attempting to bootstrap the project behind the scenes, but that bootstrapping process is not being reported to RStudio? What happens if you launch an R process in that folder from the terminal?

**Comment by jjallaire** [comment]
Kevin, I believe that what you are suggesting is the case. I've seen this
before as well where when upgrading to a new version of R the packrat
bootstrap blocks the world in the IDE with no user feedback.

I feel like we designed the IDE integration to prevent this, but I've
definitely seen it. Perhaps something changed in packrat or the IDE at some
point to make this work differently?

**Comment by kevinushey** [comment]
Right, packrat has some logic that should handle this in the init script: https://github.com/rstudio/packrat/blob/master/inst/resources/init.R#L54-L63

RStudio should be calling that on deferred init here: https://github.com/rstudio/rstudio/blob/master/src/cpp/session/SessionMain.cpp#L1944-L1945

I'll see if we're for some reason not going through that code path (perhaps packrat is bootstrapping over-eagerly for some reason)

**Comment by kevinushey** [comment]
It looks like packrat is doing the right thing here (tested with latest packrat + latest RStudio release on Windows and OS X); e.g. this is what I see in the IDE when attempting to open a packrat project with no private library:

![screen shot 2016-05-06 at 1 09 55 pm](https://cloud.githubusercontent.com/assets/1976582/15084956/dc485892-138b-11e6-80e8-3508aa298622.png)

**Comment by jjallaire** [comment]
How about for an R upgrade under an existing packified project with a large
number of packages?

**Comment by grrothman** [comment]
I'm running RStudio Version 0.99.896 on OSX 10.11.4.

I also want to add that I can initiate packrat in a new project just fine, the step which you seem to have shown in your screenshot. The problem that I'm running into is that when I create a new piece of source code, or try to open an existing .R script, the RStudio instance freezes up.

**Comment by kevinushey** [comment]
Interesting; thanks. Does this only occur in packrat-managed projects? I wonder if some of the indexing RStudio is trying to perform behind the scenes is barfing for some reason?

**Comment by grrothman** [comment]
Yes, I could open projects that were not running packrat and open old source code or create new source code without a problem. It was only when I checked the 'use packrat with this project' box in the Create New Project process, or ran `packrat::init()` that I was unable to open any code.

To further clarify, when I checked the packrat box during project setup, the RStudio session would immediately crap out. The entire window was white with none of the pane splits or menus, and basically froze there. If I didn't check the box, the project would load OK, and I could then load packrat successfully via `packrat::init()`. However, at this point, if I tried to open old code, after selecting a file to load the whole session window went white, and if I tried to create a new source code doc (cmd+shift+N), only the source code pane would go white but the whole window froze up.

**Comment by kevinushey** [comment]
Are you able to run `clang` in the console? E.g. what's the output of the following commands:

```
clang -v
xcode-select -p
```

Similarly, what happens if you try to run `.rs.canBuildCpp()` in the IDE?

**Comment by grrothman** [comment]
I've posted a video on youtube walking through the steps that I've tried and specifically, the various things that fail when packrat is enabled on an RStudio project running R v3.3.0 and packrat v0.4.7-1. I think this visual will be more helpful than a verbal description. There are additional notes in the video description with timestamps describing what I am attempting and what's failing. Please let me know if there is any additional information that is missing or would be helpful for your troubleshooting.
https://www.youtube.com/watch?v=Fmn1qcpJHkM

**Comment by kevinushey** [comment]
Hi Gabe,

Thanks for posting the video -- makes it a lot more clear what exactly is going on. It looks like the RStudio is hanging when attempting to initialize the R session; odds are it's during the packrat initialization. Can you double check a couple other things?

- Just to confirm, do you see this hang if you try to use packrat + R outside of RStudio (e.g. just in the terminal)? I'm guessing not, but would be worth confirming.

- What happens if you remove the `.Rprofile` file within the project directory before launching RStudio? (Ie, initialize packrat from the R console, then attempt to open that project in RStudio). You could also try removing / tinkering the file generated at `packrat/init.R`; e.g. try adding a `browser()` command and stepping through the code.

If you have time, can you test with the old release + current preview of RStudio? You can download them here:

https://s3.amazonaws.com/rstudio-dailybuilds/RStudio-0.98.1091.dmg (note: doesn't have packrat integration)
https://www.rstudio.com/products/rstudio/download/preview/

Thanks for digging in!

**Comment by grrothman** [comment]
The dependency warnings go on for a while. The console doesn't hang up, but it also never gets to the point where it restarts the R session.
![screen shot 2016-05-11 at 2 01 42 pm](https://cloud.githubusercontent.com/assets/12799297/15191069/f12a3f3e-1780-11e6-8de7-6cc3cad63cec.png)

**Comment by grrothman** [comment]
Woah ok, on to something here. So I created the project, initialized packrat in the terminal, and then force quit out of the RStudio session when it hung up. Then I deleted the `.Rprofile` file from the project folder and relaunched the project (just double-clicking `testing123.RProj`), and I was able to create new source code. I was also able to load a library, and execute packrat functions.

![screen shot 2016-05-11 at 2 12 53 pm](https://cloud.githubusercontent.com/assets/12799297/15191420/a244fab0-1782-11e6-9b60-1f68a651bb82.png)

**Comment by kevinushey** [comment]
Interesting, it's definitely got something to do with the RStudio packrat initialization then.

What happens if you call `source("packrat/init.R")` from that project now?

**Comment by grrothman** [comment]
![screen shot 2016-05-11 at 3 04 22 pm](https://cloud.githubusercontent.com/assets/12799297/15192959/aac42be6-1789-11e6-9325-fbc26e5d8430.png)

**Comment by kevinushey** [comment]
That would imply that RStudio's packrat-related initialization worked as expected. I wonder why it's not working on startup, then?

Can you try debugging / tracing the code within `packrat/init.R`? Note that this gets called when the R process is launched (through the `.Rprofile`), and there's logic within to defer initialization for RStudio. I'm curious if execution of that code is completing successfully.

**Comment by grrothman** [comment]
I copied the code from `packrat/init.R` and ran it in the RStudio session, it looks like there are two parts that throw errors

```
> libDir <- if (exists(".rs.getProjectDirectory"))
+     getPackratLibDir(.rs.getProjectDirectory())
>   else
Error: unexpected 'else' in "  else"
>     getPackratLibDir()
[1] "packrat/lib/x86_64-apple-darwin13.4.0/3.3.0"
```

and

```
>   if (suppressWarnings(requireNamespace("packrat", quietly = TRUE, lib.loc = libDir))) {
+     
+     # Check 'print.banner.on.startup' -- when NA and RStudio, don't print
+     print.banner <- packrat::get_opts("print.banner.on.startup")
+     if (print.banner == "auto" && is.na(Sys.getenv("RSTUDIO", unset = NA))) {
+       print.banner <- TRUE
+     } else {
+       print.banner <- FALSE
+     }
+     return(packrat::on(print.banner = print.banner))
+   }
Error: no function to return from, jumping to top level
```
but I'm guessing the second one is because I ran it outside of the `local()` wrapping.
